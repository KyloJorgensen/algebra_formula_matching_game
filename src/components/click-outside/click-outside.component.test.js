import React, { Component } from 'react';
import { shallow, mount } from 'enzyme';
import { ClickOutsideContext } from './click-outside.context.js';
import ClickOutside from './click-outside.component.js';

describe('Render ClickOutside', () => {
	it('should match the snapshot', () => {
		const wrapper = shallow(<ClickOutside/>);
		expect(wrapper).toMatchSnapshot();
	});
});


describe('interaction', () => {
	class ComponentDidMount extends Component { 
		constructor(props) {
			super(props);
		}

		componentDidMount() {
			this.props.addClickOutsideEventListener('uniqueTag', this.props.mockFunction);
		}

		componentWillUnmount() {
			this.props.removeClickOutsideEventListener('uniqueTag');
		}

		render () {
			return (<button></button>)
		}
	}
	ComponentDidMount.defaultProps = {
		addClickOutsideEventListener: () => {},
		mockFunction: () => {},
	};

	class UnmountButton extends Component { 
		constructor(props) {
			super(props);
			this.state = {
				mount: true,
			};
			this.handleClick = event => {
				const _state = this.state;
				_state.mount = false;
				this.setState(_state);
			};
		}

		render () {
			return (
				<div>
					<button className="unmount-button" onClick={this.handleClick}></button>
					{this.state.mount ? this.props.children : ''}
				</div>
			);
		}
	}
	let wrapper;

	beforeEach(() => {


	})

	it('child component able to add a clickOutsideListener', () => {
		const mockAddClickOutsideEventListener = jest.fn().mockReturnValue({_state: {'uniqueTag': () => {}}, added: true});;
		wrapper = mount(
			<ClickOutside addClickOutsideEventListener={mockAddClickOutsideEventListener} >
				<div>
					<b></b>
					<ClickOutsideContext.Consumer>
						{context => {
							return (<ComponentDidMount addClickOutsideEventListener={context.addClickOutsideEventListener} />);
						}}
					</ClickOutsideContext.Consumer>
				</div>
			</ClickOutside>
		);

		const mockCallLength = mockAddClickOutsideEventListener.mock.calls.length;
		expect(mockCallLength).toEqual(1);
	});

	it('onClick should run all of the clickOutsideListeners', () => {
		const mockFunction = jest.fn().mockImplementation();
		wrapper = mount(
			<ClickOutside >
				<div>
					<b></b>
					<ClickOutsideContext.Consumer>
						{context => {
							return (<ComponentDidMount mockFunction={mockFunction} addClickOutsideEventListener={context.addClickOutsideEventListener} />);
						}}
					</ClickOutsideContext.Consumer>
				</div>
			</ClickOutside>
		);
		wrapper.find('b').simulate('click');

		const mockCallLength = mockFunction.mock.calls.length;
		expect(mockCallLength).toEqual(1);
	});

	it('child component able to remove a clickOutsideListener', () => {
		const mockFunction = jest.fn().mockImplementation();
		const mockRemoveClickOutsideEventListener = jest.fn().mockImplementation();
		wrapper = mount(
			<ClickOutside >
				<div>
					<b></b>
					<UnmountButton>
						<ClickOutsideContext.Consumer>
							{context => {
								let props = {
									mockFunction: mockFunction,
									addClickOutsideEventListener: context.addClickOutsideEventListener,
									removeClickOutsideEventListener: context.removeClickOutsideEventListener,
								}
								return (<ComponentDidMount {...props} />);
							}}
						</ClickOutsideContext.Consumer>
					</UnmountButton>
				</div>
			</ClickOutside>
		);
		wrapper.find('button.unmount-button').simulate('click');

		let mockCallLength = mockFunction.mock.calls.length;
		expect(mockCallLength).toEqual(1);

		wrapper.find('b').simulate('click');
		mockCallLength = mockFunction.mock.calls.length;
		expect(mockCallLength).toEqual(1);
	});

});