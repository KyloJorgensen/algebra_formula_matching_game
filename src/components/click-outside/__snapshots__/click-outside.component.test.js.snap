// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ClickOutside should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ClickOutside
    addClickOutsideEventListener={[Function]}
    removeClickOutsideEventListener={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <div
        onClick={[Function]}
      />,
      "value": Object {
        "addClickOutsideEventListener": [Function],
        "removeClickOutsideEventListener": [Function],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": undefined,
        "onClick": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": "div",
    },
    "type": Object {
      "$$typeof": Symbol(react.provider),
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": [Circular],
        "_calculateChangedBits": null,
        "_changedBits": 0,
        "_currentRenderer": null,
        "_currentValue": Object {
          "addClickOutsideEventListener": [Function],
          "removeClickOutsideEventListener": [Function],
        },
        "_defaultValue": Object {
          "addClickOutsideEventListener": [Function],
          "removeClickOutsideEventListener": [Function],
        },
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <div
          onClick={[Function]}
        />,
        "value": Object {
          "addClickOutsideEventListener": [Function],
          "removeClickOutsideEventListener": [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      "type": Object {
        "$$typeof": Symbol(react.provider),
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": [Circular],
          "_calculateChangedBits": null,
          "_changedBits": 0,
          "_currentRenderer": null,
          "_currentValue": Object {
            "addClickOutsideEventListener": [Function],
            "removeClickOutsideEventListener": [Function],
          },
          "_defaultValue": Object {
            "addClickOutsideEventListener": [Function],
            "removeClickOutsideEventListener": [Function],
          },
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
